react/src/hooks/useColumnOrderCache.tsx

```diff
@@ -0,0 +1,73 @@
+import React, { useEffect, useState } from "react";
+import { MaterialTableProps } from "@material-table/core";
+import { getTableColumnOrder } from "../functions";
+
+/**
+ * Handle updating a given material-table with localStorage-cached column order.
+ *
+ * If any dependencies are specified, ensure that they are all true before
+ * loading column order from cache. Dependencies include any other state variables
+ * or queries that must be completed for the table to render properly on first try.
+ *
+ * Return a function for updating the cache on column drag.
+ */
+export function useColumnOrderCache(
+    tableRef: React.MutableRefObject<any>,
+    cacheKey: string,
+    dependencies?: boolean[]
+) {
+    // ensures that the side effect only occurs once
+    const [applied, setApplied] = useState(false);
+
+    const handleOrderChange: MaterialTableProps<object>["onColumnDragged"] = () => {
+        // source and dest are just the visible indexes
+        // not the true indexes that take hidden columns into account
+        // so we just dig the the new column orders out of the table's data manager
+        const columnOrderRecord = getTableColumnOrder(tableRef);
+        if (columnOrderRecord !== null) {
+            localStorage.setItem(cacheKey, JSON.stringify(columnOrderRecord));
+        } else {
+            console.error(`Failed to cache column order for ${cacheKey}.`);
+        }
+    };
+
+    useEffect(() => {
+        if (
+            tableRef.current &&
+            (dependencies === undefined || dependencies.find(dep => !dep) === undefined) &&
```

Comment:
I think this may have unintended consequences, since `[true,2,undefined]` would also return `undefined` (if `undefined` is a possible dep value).